File: core/services/push.py
import json
import logging

from django.conf import settings
from django.utils import timezone

from core.models import Device, PushSubscription

logger = logging.getLogger(__name__)

try:
    from pywebpush import WebPushException, webpush
except ImportError:  # pragma: no cover - optional at runtime
    WebPushException = Exception
    webpush = None


def push_is_configured() -> bool:
    return bool(
        settings.LINKHOP_WEBPUSH_VAPID_PUBLIC_KEY
        and settings.LINKHOP_WEBPUSH_VAPID_PRIVATE_KEY
        and webpush is not None
    )


def get_public_push_config() -> dict:
    return {
        "supported": push_is_configured(),
        "vapid_public_key": settings.LINKHOP_WEBPUSH_VAPID_PUBLIC_KEY if push_is_configured() else "",
    }


def upsert_push_subscription(
    *,
    device: Device,
    endpoint: str,
    p256dh: str,
    auth_secret: str,
    user_agent: str = "",
) -> PushSubscription:
    subscription, _ = PushSubscription.objects.update_or_create(
        endpoint=endpoint,
        defaults={
            "device": device,
            "p256dh": p256dh,
            "auth_secret": auth_secret,
            "user_agent": user_agent[:255],
            "is_active": True,
            "last_error": "",
        },
    )
    return subscription


def deactivate_push_subscription(*, device: Device, endpoint: str) -> int:
    return PushSubscription.objects.filter(
        device=device,
        endpoint=endpoint,
        is_active=True,
    ).update(
        is_active=False,
        updated_at=timezone.now(),
    )


def _send_to_subscriptions(*, device: Device, payload: str, subscriptions) -> dict:
    """Send a push payload to a queryset of subscriptions. Returns delivery stats."""
    delivered = 0
    total = 0

    for subscription in subscriptions:
        total += 1
        try:
            webpush(
                subscription_info={
                    "endpoint": subscription.endpoint,
                    "keys": {
                        "p256dh": subscription.p256dh,
                        "auth": subscription.auth_secret,
                    },
                },
                data=payload,
                vapid_private_key=settings.LINKHOP_WEBPUSH_VAPID_PRIVATE_KEY,
                vapid_claims={"sub": settings.LINKHOP_WEBPUSH_VAPID_SUBJECT},
            )
        except WebPushException as exc:
            status_code = getattr(getattr(exc, "response", None), "status_code", None)
            subscription.last_failure_at = timezone.now()
            subscription.last_error = str(exc)[:255]
            if status_code in {404, 410}:
                subscription.is_active = False
            subscription.save(
                update_fields=[
                    "last_failure_at",
                    "last_error",
                    "is_active",
                    "updated_at",
                ]
            )
            logger.warning("Push delivery failed for %s: %s", subscription.endpoint, exc)
            continue

        delivered += 1
        now = timezone.now()
        subscription.last_success_at = now
        subscription.last_error = ""
        subscription.save(update_fields=["last_success_at", "last_error", "updated_at"])
        Device.objects.filter(id=device.id).update(last_push_at=now)

    return {"delivered": delivered, "total": total}


def relay_push_message(
    *,
    device: Device,
    message_id: str,
    message_type: str,
    body: str,
    sender_name: str,
    recipient_device_id: str,
    created_at: str,
    is_test: bool = False,
) -> dict:
    """Relay a message to a device via Web Push. Returns delivery stats."""
    if not push_is_configured():
        return {"delivered": 0, "total": 0}

    payload = json.dumps(
        {
            "message_id": message_id,
            "type": message_type,
            "body": body,
            "sender": sender_name,
            "recipient_device_id": recipient_device_id,
            "created_at": created_at,
            **({"test": True} if is_test else {}),
        }
    )

    subscriptions = PushSubscription.objects.filter(device=device, is_active=True)
    return _send_to_subscriptions(device=device, payload=payload, subscriptions=subscriptions)

File: linkhop-ts/src/services/push.ts
import type { Database } from '@db/sqlite';
import webpush from 'web-push';
import type {
  AppConfig,
  DeviceRecord,
  PushSubscriptionRecord,
} from '../types.ts';

export function pushIsConfigured(config: AppConfig): boolean {
  return Boolean(config.vapidPublicKey && config.vapidPrivateKey);
}

export function getPublicPushConfig(config: AppConfig): {
  supported: boolean;
  vapid_public_key: string;
} {
  return {
    supported: pushIsConfigured(config),
    vapid_public_key: pushIsConfigured(config) ? config.vapidPublicKey : '',
  };
}

export function upsertPushSubscription(
  db: Database,
  input: {
    device: DeviceRecord;
    endpoint: string;
    p256dh: string;
    authSecret: string;
    clientType?: string;
    userAgent?: string;
  },
): void {
  const statement = db.prepare(`
    INSERT INTO push_subscriptions (
      id, device_id, endpoint, p256dh, auth_secret, client_type, user_agent, is_active, last_error
    ) VALUES (?, ?, ?, ?, ?, ?, ?, 1, NULL)
    ON CONFLICT(endpoint) DO UPDATE SET
      device_id = excluded.device_id,
      p256dh = excluded.p256dh,
      auth_secret = excluded.auth_secret,
      client_type = excluded.client_type,
      user_agent = excluded.user_agent,
      is_active = 1,
      last_error = NULL
  `);

  try {
    statement.run([
      crypto.randomUUID(),
      input.device.id,
      input.endpoint,
      input.p256dh,
      input.authSecret,
      (input.clientType || '').slice(0, 20) || null,
      (input.userAgent || '').slice(0, 255) || null,
    ]);
  } finally {
    statement.finalize();
  }
}

export function deactivatePushSubscription(
  db: Database,
  input: { deviceId: string; endpoint: string },
): void {
  const statement = db.prepare(`
    UPDATE push_subscriptions
    SET is_active = 0, last_failure_at = datetime('now')
    WHERE device_id = ? AND endpoint = ? AND is_active = 1
  `);

  try {
    statement.run([input.deviceId, input.endpoint]);
  } finally {
    statement.finalize();
  }
}

function listActiveSubscriptions(
  db: Database,
  deviceId: string,
): PushSubscriptionRecord[] {
  const statement = db.prepare(`
    SELECT * FROM push_subscriptions
    WHERE device_id = ? AND is_active = 1
    ORDER BY created_at ASC
  `);

  try {
    const subscriptions = statement.all([deviceId]) as PushSubscriptionRecord[];
    const extensionSubscriptions = subscriptions.filter((subscription) =>
      subscription.client_type === 'extension'
    );
    return extensionSubscriptions.length
      ? extensionSubscriptions
      : subscriptions;
  } finally {
    statement.finalize();
  }
}

function markPushSuccess(
  db: Database,
  deviceId: string,
  subscriptionId: string,
): void {
  const subscriptionStatement = db.prepare(`
    UPDATE push_subscriptions
    SET last_success_at = datetime('now'), last_error = NULL
    WHERE id = ?
  `);
  const deviceStatement = db.prepare(`
    UPDATE devices
    SET last_push_at = datetime('now')
    WHERE id = ?
  `);

  try {
    subscriptionStatement.run([subscriptionId]);
    deviceStatement.run([deviceId]);
  } finally {
    subscriptionStatement.finalize();
    deviceStatement.finalize();
  }
}

function markPushFailure(
  db: Database,
  subscriptionId: string,
  message: string,
  deactivate = false,
): void {
  const statement = db.prepare(`
    UPDATE push_subscriptions
    SET
      last_failure_at = datetime('now'),
      last_error = ?,
      is_active = CASE WHEN ? THEN 0 ELSE is_active END
    WHERE id = ?
  `);

  try {
    statement.run([message.slice(0, 255), deactivate ? 1 : 0, subscriptionId]);
  } finally {
    statement.finalize();
  }
}

export async function relayPushMessage(
  db: Database,
  config: AppConfig,
  input: {
    device: DeviceRecord;
    messageId: string;
    messageType: string;
    body: string;
    senderName: string;
    recipientDeviceId: string;
    createdAt: string;
    isTest?: boolean;
  },
): Promise<{ delivered: number; total: number }> {
  if (!pushIsConfigured(config)) {
    return { delivered: 0, total: 0 };
  }

  const subscriptions = listActiveSubscriptions(db, input.device.id);
  if (!subscriptions.length) {
    return { delivered: 0, total: 0 };
  }

  webpush.setVapidDetails(
    config.vapidSubject,
    config.vapidPublicKey,
    config.vapidPrivateKey,
  );

  const payload = JSON.stringify({
    message_id: input.messageId,
    type: input.messageType,
    body: input.body,
    sender: input.senderName,
    recipient_device_id: input.recipientDeviceId,
    created_at: input.createdAt,
    ...(input.isTest ? { test: true } : {}),
  });

  let delivered = 0;

  for (const subscription of subscriptions) {
    try {
      await webpush.sendNotification(
        {
          endpoint: subscription.endpoint,
          keys: {
            p256dh: subscription.p256dh,
            auth: subscription.auth_secret,
          },
        },
        payload,
      );
      delivered += 1;
      markPushSuccess(db, input.device.id, subscription.id);
    } catch (error) {
      const statusCode = error instanceof Error && 'statusCode' in error
        ? Number(error.statusCode)
        : undefined;
      markPushFailure(
        db,
        subscription.id,
        error instanceof Error ? error.message : 'Push delivery failed',
        statusCode === 404 || statusCode === 410,
      );
    }
  }

  return { delivered, total: subscriptions.length };
}
